### PR TITLE
PEP 740 blog post

### LINKED ISSUE #15871: [Roadmap] Implement PEP 740
[PEP 740](https://peps.python.org/pep-0740/) is in a final but not yet approved state. This issue is intended to lay out the dependencies/subcomponents of its implementing, including things that can be done in a preliminary manner.

## Index side

- [ ] Support the `attestations` form field in the current legacy upload endpoint
- [ ] Parse each received attestation from `attestations` and verify each against the uploading trusted publisher; fail the file upload if any attestations are invalid
  * This should be done in as generic a manner as possible, to ensure that future extension to email and key identities is possible.
- [ ] Persist verified attestations and expose them via the PEP 503 and Simple JSON API indices (via provenance objects, as specified in PEP 740)

## Uploader/publish side

- [ ] Support attestation generation in `gh-action-pypi-publish`
  * Concretely, this means `gh-action-pypi-publish` should use `sigstore-python` to sign the attestation payload defined in PEP 740
  * The resulting bundle should be transformed into a PEP 740 attestation object
- [ ] Support attestation uploading in `twine`: https://github.com/pypa/twine/issues/1094
- [ ] Once `twine` supports uploading, `gh-action-pypi-publish` should use that support to actually upload the attestation objects it generates above

CC @di @webknjaz @facutuesca for visibility

### PR DESCRIPTION
**Pull request by di**
Fixes #15871.

Needs:

- [x] Update ToB blog post permalink

### CODE REVIEW

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  case of an upload of a project built with GitHub Actions, PyPI's support for
+  digital attestations defines a strong and verifiable association between a file
+  on PyPI and the source repository, workflow, and even the commit hash that
+  produced and uploaded the file.
+* **Attestations are verified when uploaded, and must be verifiable to be uploaded:**
+  Upon review of the state of historical PGP signatures published to PyPI, [we
+  found that many signatures were not verifiable], either by PyPI or by end
+  users. With support for PEP 740, PyPI only permits attestations with a
+  verifiable signature to be uploaded and redistributed by the index. This
+  ensures that all attestations are verifiable and useful for all PyPI users.
```
Comment (haydentherapper) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:49]: Could also mention the transparency and auditability of attestations to mitigate against compromise of the repository?

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+3. use the [pypa/gh-action-pypi-publish] action to publish.
+
+Support for automatic attestation generation and publication from other Trusted
+Publisher environments [is planned]. While not recommended, maintainers can also
+[manually generate and publish attestations].
+
+### Acknowledgements
+
+Support for work on PEP 740's authoring and design was provided by the
+[Soverign Tech Fund] and the Google Open Source Security Team.
```
Comment (haydentherapper) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:87]: I believe it's "Sovereign Tech Agency" now (cc @woodruffw)

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+### Get started today
+
+The generation and publication of attestations happens by default, and no
+changes are necessary for projects that meet all of these conditions:
+
+1. publish from GitHub Actions;
+2. via [Trusted Publishing]; and
+3. use the [pypa/gh-action-pypi-publish] action to publish.
+
+Support for automatic attestation generation and publication from other Trusted
```
Comment (haydentherapper) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:80]: While it's linked, should we explicitly mention that attestations can be generated from other CI providers that are supported by trusted publishing?

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+### Why not plain signatures?
+
+PyPI's support for digital attestations has three key advantages over regular
+cryptographic signatures, such as those provided by PGP:
+
+* **Attestations are signed by an identity, not a key pair:**
+  Similar to our recent support for Trusted Publishing, PyPI's support for
+  digital attestations relies upon Open ID Connect (OIDC) identities. By signing
+  attestations with identities, and not a public/private key pair, we mitigate the
```
Comment (haydentherapper) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:35]:
```suggestion
  attestations with identities, popularized by the [Sigstore project](https://www.sigstore.dev/), and not a public/private key pair, we mitigate the
```

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md (hunk 3 above)
Comment (di) [reply docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:80]: Only GitHub is currently supported: https://docs.pypi.org/attestations/producing-attestations/#:~:text=Attestations%20are%20currently%20only%20supported%20when%20uploading%20with%20Trusted%20Publishing%2C%20and%20currently%20only%20with%20GitHub%2Dbased%20Trusted%20Publishers

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+[PEP 740]: https://peps.python.org/pep-0740/
+[Trusted Publishing]: https://docs.pypi.org/trusted-publishers/
+[deprecation and removal of PGP signatures]: https://blog.pypi.org/posts/2023-05-23-removing-pgp/
+[pypa/gh-action-pypi-publish]: https://github.com/pypa/gh-action-pypi-publish
+[Attestations: a new generation of signatures on PyPI]: https://blog.trailofbits.com/2024/11/14/attestations-a-new-generation-of-signatures-on-pypi
+[Integrity API]: https://docs.pypi.org/api/integrity/
+[we announced last year]: 2023-04-20-introducing-trusted-publishers.md
+[we found that many signatures were not verifiable]: https://blog.yossarian.net/2023/05/21/PGP-signatures-on-PyPI-worse-than-useless
+[manually generate and publish attestations]: https://docs.pypi.org/attestations/producing-attestations/#the-manual-way
+[Soverign Tech Fund]: https://www.sovereign.tech/tech/python-package-index
```
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:111]:
```suggestion
[Sovereign Tech Agency]: https://www.sovereign.tech/tech/python-package-index
```

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md (hunk 2 above)
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:87]:
```suggestion
[Sovereign Tech Agency] and the Google Open Source Security Team.
```

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  digital attestations relies upon Open ID Connect (OIDC) identities. By signing
+  attestations with identities, and not a public/private key pair, we mitigate the
+  potential for an individual's key loss or compromise, one of the most common failure cases for
+  PGP signing.
+* **Attestations provide a verifiable link to an upstream source repository:**
+  By signing with the identity of the upstream source repository, such as in the
+  case of an upload of a project built with GitHub Actions, PyPI's support for
+  digital attestations defines a strong and verifiable association between a file
+  on PyPI and the source repository, workflow, and even the commit hash that
+  produced and uploaded the file.
```
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:43]:
```suggestion
  on PyPI and the source repository, workflow, and even the commit hash that
  produced and uploaded the file. Additionally, publishing attestations to a
  transparency log helps mitigate against both compromise of PyPI and compromise
  of the projects themselves.
```

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  - security
+  - oidc
+---
+
+PyPI package maintainers can now publish signed digital attestations when
+publishing, in order to further increase trust in the supply-chain security of
+their projects. Additionally, a new API is available for consumers and
+installers to verify published attestations.
+
+Many projects have already begun publishing attestations, with more than 10,000
```
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:18]:
```suggestion
Many projects have already begun publishing attestations, with more than 20,000
```

**Review by ewdurbin — APPROVED** [review]

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+### Acknowledgements
+
+Support for work on PEP 740's authoring and design was provided by the
+[Sovereign Tech Agency] and the Google Open Source Security Team.
+
+Funding for the implementation of PEP 740 was provided by the Google Open
+Source Security Team, and much of the development work on PyPI and related
+tools was performed by [Trail of Bits], with special thanks to contributors
+[William Woodruff] and [Facundo Tuesca].
+
```
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:95]:
```suggestion

Thanks to the the [Sigstore project] for their work popularizing identity-based signing, hosting a public-good transparency log, and continued support of the [Python client for Sigstore].

```

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md (hunk 4 above)
Comment (di) [reply docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:35]: 261b6709f3d0800f207d5b44fa32edc990f637e0

File: docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+---
+
+_Dustin Ingram is a maintainer of the Python Package Index._
+
+[PEP 740]: https://peps.python.org/pep-0740/
+[Trusted Publishing]: https://docs.pypi.org/trusted-publishers/
+[deprecation and removal of PGP signatures]: https://blog.pypi.org/posts/2023-05-23-removing-pgp/
+[pypa/gh-action-pypi-publish]: https://github.com/pypa/gh-action-pypi-publish
+[Attestations: a new generation of signatures on PyPI]: https://blog.trailofbits.com/2024/11/14/attestations-a-new-generation-of-signatures-on-pypi
```
Comment (di) [inline docs/blog/posts/2024-11-14-pypi-now-supports-digital-attestations.md:110]:
```suggestion
[Attestations: a new generation of signatures on PyPI]: https://blog.trailofbits.com/2024/11/14/attestations-a-new-generation-of-signatures-on-pypi/
```

**Pull request merged**